contracts/src/gateway/BaseForeignGateway.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "../arbitration/IArbitrable.sol";
+import "../bridge/IL1Bridge.sol";
+
+import "./interfaces/IHomeGateway.sol";
+import "./interfaces/IForeignGateway.sol";
+
+abstract contract BaseForeignGateway is IL1Bridge, IForeignGateway {
+    // @dev Note the disputeID needs to start from one as
+    // the KlerosV1 proxy governor depends on this implementation.
+    uint256 internal localDisputeID = 1;
+
+    // For now this is just a constant, but we'd probably need to
+    // implement the same arbitrationCost calculation code we'll have
+    // in the V2 court.
+    uint256 internal internalArbitrationCost;
+
+    struct DisputeData {
+        uint256 id;
+        address arbitrable;
+    }
+    mapping(uint256 => bytes32) public disputeIDtoHash;
+    mapping(bytes32 => DisputeData) public disputeHashtoDisputeData;
+
+    IHomeGateway public homeGateway;
+    uint256 public chainID;
+
+    modifier onlyFromL2() {
+        this.onlyAuthorized();
+        _;
+    }
+
+    constructor(uint256 _arbitrationCost, IHomeGateway _homeGateway) {
+        internalArbitrationCost = _arbitrationCost;
+        homeGateway = _homeGateway;
+
+        uint256 id;
+        assembly {
+            id := chainid()
+        }
+        chainID = id;
+    }
+
+    function createDispute(uint256 _choices, bytes calldata _extraData) external payable returns (uint256 disputeID) {
+        require(msg.value >= arbitrationCost(_extraData), "Not paid enough for arbitration");
+
+        disputeID = localDisputeID++;
+        bytes32 disputeHash = keccak256(
+            abi.encodePacked(
+                chainID,
+                blockhash(block.number - 1),
+                "createDispute",
+                disputeID,
+                arbitrationCost(_extraData),
+                _extraData
```

Comment:
Let's add the msg.sender to the disputeHash, as the gateway may receive disputes from other contracts later.